src/test/java/com/squareup/javapoet/TypeNameTest.java

```diff
@@ -32,6 +32,34 @@
   protected <E extends Enum<E>> E generic(E[] values) {
     return values[0];
   }
+  
+  protected static class TestGeneric<T> {
+    class Inner {}
+
+    class InnerGeneric<T2> {}
+
+    static class NestedNonGeneric {}
+  }
+
+  protected static TestGeneric<String>.Inner testGenericStringInner() {
+    return null;
+  }
+
+  protected static TestGeneric<Integer>.Inner testGenericIntInner() {
+    return null;
+  }
+
+  protected static TestGeneric<String>.InnerGeneric<String> testGenericInnerString() {
```

Comment:
Would it be prudent to use different type parameters? So if you have a bug that mixes them up, you catch it?